target-chains/near/receiver/src/governance.rs

```diff
@@ -34,162 +33,218 @@ use {
+                        let (input, bytes) = take(32usize)(input)?;
+                        let chain = Chain::from(WormholeChain::from(chain));
+                        let mut emitter = [0u8; 32];
+                        emitter.copy_from_slice(bytes);
+                        Ok((input, Source { chain, emitter }))
+                    }))(input)?;
+                    GovernanceAction::SetDataSources { data_sources }
+                }
 
-            ActionId::SetGovernanceSource => {
-                let mut emitter = [0u8; 32];
-                cursor.read_exact(&mut emitter)?;
-                Self::SetGovernanceSource(Source {
-                    emitter,
-                    pyth_emitter_chain: Chain(cursor.read_u16::<BigEndian>()?),
-                })
-            }
+                GovernanceActionId::SetFee => {
+                    let (_input, (val, expo)) = all_consuming(|input| {
+                        let (input, val) = be_u64(input)?;
+                        let (input, expo) = be_u64(input)?;
+                        Ok((input, (val, expo)))
+                    })(input)?;
+                    GovernanceAction::SetFee { base: val, expo }
+                }
 
-            ActionId::SetStalePriceThreshold => {
-                let stale_price_threshold = cursor.read_u64::<BigEndian>()?;
-                Self::SetStalePriceThreshold(stale_price_threshold)
-            }
+                GovernanceActionId::SetValidPeriod => {
+                    let (_input, valid_seconds) = all_consuming(be_u64)(input)?;
+                    GovernanceAction::SetValidPeriod { valid_seconds }
+                }
 
-            ActionId::SetUpdateFee => {
-                let update_fee = cursor.read_u64::<BigEndian>()?;
-                Self::SetUpdateFee(update_fee)
-            }
+                GovernanceActionId::RequestGovernanceDataSourceTransfer => {
+                    let (_input, governance_data_source_index) = all_consuming(be_u32)(input)?;
+                    GovernanceAction::RequestGovernanceDataSourceTransfer {
+                        governance_data_source_index,
+                    }
+                }
+            },
         })
     }
 
-    pub fn serialize(&self) -> Vec<u8> {
-        let mut data = Vec::new();
-        let magic = u32::from_le_bytes(GOVERNANCE_MAGIC);
-        data.write_u32::<BigEndian>(magic).unwrap();
-        data.push(GOVERNANCE_MODULE);
-        data.push(self.id() as u8);
-        data.extend_from_slice(&0u16.to_le_bytes());
-
-        match self {
-            Self::ContractUpgrade(hash) => {
-                data.extend_from_slice(hash);
+    /// Implements a `serialize` method for the `GovernanceAction` enum. The `nom` library doesn't
+    /// provide serialization but serialization is a safer operation, so we can just use a simple
+    /// push buffer to serialize.
+    pub fn serialize(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = Vec::new();
```

Comment:
Dumb question, but when do we need to serialize governance actions?

Reply:
I needed a nice way to get these in tests, I can mark this `#[cfg(test)]` but I figured the utility might be worth leaving it available within the library.

Reply:
Side point - for paranoia, it's probably good to have the test VAAs be generated completely orthogonally to the nom logic, so that if there is a bug in serialize/deserialize it'll get caught. You can use the [`xc-governance-sdk-js`](https://github.com/pyth-network/pyth-crosschain/tree/main/third_party/pyth/xc-governance-sdk-js) package to generate these test VAA's outside of the NEAR code.